include/ada/urlpattern.h

```diff
@@ -0,0 +1,51 @@
+#ifndef ADA_URL_PATTERN_H
+#define ADA_URL_PATTERN_H
+
+#include "ada/common_defs.h"
+
+#include "ada/urlpattern_base.h"
+
+#include <unordered_map>
+#include <string_view>
+#include <optional>
+
+namespace ada::urlpattern {
+struct urlpattern_component_result {
+  std::string_view input;
+  std::unordered_map<std::string_view, std::optional<std::string_view>> groups;
```

Comment:
@lemire How's the performance comparison of using std::optional in here versus std::variant<std::nullopt, std::string_view>?

Reply:
I don't expect it matters much.